### ISSUE TITLE
API for "Request Address"

### ISSUE THREAD

**Issue by Nutomic** [schildbach/bitcoin-wallet#212]
I am developing a decentralized, encrypted instant messenger, [ensichat](https://github.com/Nutomic/ensichat), and thought it would be cool to send Bitcoins in the chat.

As I understand, your API only supports sending coins to a given address. What I need is an API for "Request Address", so I know where to send coins. Additionally, there should be a callback when coins are received on that address, so I can mark the payment as received in my app.

Do you think this makes sense, and could be implemented like this?

**Comment by schildbach** [comment]
So what you want is fetching a Bitcoin address (or a Payment Request message) from the Bitcoin Wallet app and hand it over to chat partners?

I'm not sure if this can be done in a clean and safe way. Have you considered implementing your own small in-app-wallet using bitcoinj?

**Comment by Nutomic** [comment]
Yeah exactly, and I'm not sure about the security myself (I only know Bitcoin as a user). What problems in particular do you see with fetching a Bitcoin address? Anyway, this could require a permission, or use a yes/no dialog where the user has to agree before handing out the address.

Implementing my own wallet sounds rather complicated. It also means that the user would have to transfer Bitcoins manually between different apps on his phone, and that's bad UX imo.

**Comment by schildbach** [comment]
Yes, that might work, at least if you don't need the addresses / payment requests prophylactically. That means a user must actively initiate a Bitcoin payment before the wallet is asked for the address. That should pretty much cover the first half of your requirement.

As for the second half (notification of payment), can you say how you plan to show that in your app? Will there be a list of open payments?

**Comment by Nutomic** [comment]
I thought it could work like this:
Alice clicks "send bitcoins"
  Alice's device sends "get address" message to Bob's device
  Bob's device fetches Bitcoin address from wallet and sends it back
  Alice's device displays dialog to enter amount
  Payment is displayed in the chat on Alice's device (similar to a normal text message)
Alice enters amount to send and confirms
  Bob's Bitcoin wallet notifies the messenger app of the received payment
  Payment is displayed in the chat on Bob's device

Alternatively, Alice's device could send a "payment finished" message to Bob's device, but there should be a way to verify with the wallet that this payment is valid (and maybe check number of confirmations).

**Comment by schildbach** [comment]
I assume Alice and Bob are already paired up in a chat session? (Side question: What happens if there are 3 or more people in a chat room?)

I wonder if Alice should maybe enter the desired amount upfront, before an address is being asked for. It would limit the number of round trips and also make it easier to use payment request messages rather than naked addresses. Also I wonder if it would make sense to *alternatively* let Bob request coins from Alice, similar to how you *can* use wallet apps.

The most secure validation you can get is inside the wallet app itself. It will show a system notificaton for received payments – maybe that's enough already? If not, Bob's device would need to somehow subscribe to payments to that specific payment request. This is kind of hairy right now. It's not even implemented inside Bitcoin Wallet itself, although I want to get there. It may be easier to implement in future when we have switched to a payment-centric UI (as opposed to a transaction-centric UI right now).

I just understood your chat system is decentralized and uses Bluetooth. Awesome! What kind of security does your system provide to protect e.g. against MITM? Obviously the last thing we'd want is someone to open a "Bluetooth honeypot" and replace Bitcoin addresses by his own. Are there any limits on message lengths? How do you manage identity?

**Comment by Nutomic** [comment]
Yes, they are in a chat with encrypted, signed messages. I don't have group chat implemented yet, so that wouldn't be a problem for now.

Wouldn't entering the amount first require the same amount of messages? (in both cases, only the Bitcoin address has to be fetched). But it would definitely be less user friendly, because you need to enter the amount, then confirm the payment while verifying the amount hasn't changed. Or maybe I'm understanding this wrong.

A notification would work I guess, as long as my app can set the text, so the user knows where it is coming from.

You can read the [protocol here](https://github.com/Nutomic/ensichat/blob/master/PROTOCOL.md). Each user has a key pair that's used for signing and encrypting messages (done in class [Crypto](https://github.com/Nutomic/ensichat/blob/master/app/src/main/scala/com/nutomic/ensichat/protocol/Crypto.scala)). You have to add contacts first, which works only if devices are connected directly to each other (exactly one hop distance). Then both devices show identicons of their public keys so you can visually verify them [screenshot](https://raw.githubusercontent.com/Nutomic/ensichat/master/graphics/screenshot_phone_1.png). Message length is currently limited to about 4 GB (size is stored in 4 bytes).

**Comment by schildbach** [comment]
Ok maybe we should meet and discuss details on a piece of paper. I understand we're both based in Berlin (-:

Thinking about it, maybe you're right about when to enter the amount. Probably you're using a wallet app like Bitcoin Wallet also on Alice's side and when opening the send dialog there will be a field for entering an amount. By the way, when you're going to authorize the payment on Alice's side you'll get a Payment Message which you could forward to Bob's device which can submit it to via another yet to implement Intent. That way, we could run the whole BIP70 payment protocol via your chat network. That would mean payer and payee could stay anonymous to each other to the extent that they don't even know Bluetooth MAC addresses of each other.

**Comment by Nutomic** [comment]
That would be great, I'll send you a mail :)

As I said, I don't know much about the workings of sending Bitcoins, but if you implement something, I can probably work with it. In my app, contacts would know each other (because of the inital key exchange), but you're right, they wouldn't have to be close to each other or be directly connected.

**Comment by mikehearn** [comment]
It'd be great if Bitcoin Wallet exposed an intent/binder API that let other apps request a bitcoin URI and/or PaymentRequest message. Ideally a PaymentRequest message.

**Comment by schildbach** [comment]
Ah yes, I should have updated the ticket. As a result of the meeting with Felix, I now have a branch that does this. If anyone is interested, I'm happy to mail a test version plus instructions.

**Comment by Nutomic** [comment]
What is your plan for merging this? It worked fine on my devices with testnet.

**Comment by schildbach** [comment]
Can you send me your most current test version? When I tested this last time, I was not able to pair two ensichat's.

**Comment by Nutomic** [comment]
[Download](https://dl.dropboxusercontent.com/u/22457329/ensichat-send-bitcoin.apk)

I don't think I've changed anything with the pairing. It didn't work for me a few times for no apparent reason, but an app reinstall and/or device reboot always fixed it. I would appreciate any logs if it fails reproducibly.

**Comment by iNPUTmice** [comment]
Hi,

I don't what API you two came up with a year ago but I'm also interested in something similar.
I'm thinking about a simple intent like ```CREATE_INVOICE``` that launches the bitcoin wallet asks me for a particular amount, (that amount could also be provided by an intent extra) upon hitting OK the bitcoin wallet generates a new address and returns with an bitcoin uri that includes the amount.

The App that created the intent (communication app of some kind / chat app) could then send the URI to a recipient. As soon as the bitcoin wallet receives enough confirmations for a transaction on that address it generates a broadcast intent that notifies the original app. (including the transaction number)

Also there should be some kind of VIEW_BALANCE_ON_ADDRESS or VIEW_TRANSACTIONS_FOR_ADDRESS that could be used to open the bitcoin wallet when clicking on the generated 'invoice' message.

**Comment by mahalamobile** [comment]
Hi, just adding my bit here. All im looking for is an API call to get a valid address from the wallet. This will be part of a rewards program where payouts are in BTC, so will need to call bitcoin wallet for "new" payout address. Payouts are then done via another system :) please help!

**Comment by schildbach** [comment]
@mahalamobile @iNPUTmice Can you look at the https://github.com/bitcoin-wallet/bitcoin-wallet/tree/fetch-payment-request branch to see if it fits your needs?

**Comment by mahalamobile** [comment]
yup

i think the key here is this,

final Address *address *= wallet.freshReceiveAddress();

So here is the process flow if i have it correctly?

open requesting app, for example it displays balance BTC 0.001, then a
cashout button (to mobile wallet)
will it be able to get address without launching the btc app? if not then
user will be shown "do you want to receive payment" dialog, if accepts then
"callback" to originating app with the *address *? Do you have to specify
the amount?

cool

**Comment by schildbach** [comment]
The app will be automatically launched by the intent you send to it. The user will only see the confirmation dialog. The calling app will receive the payment request via "intent result" (a standard mechanism for Android apps).

Note a payment request contains a script to pay to, rather than a Bitcoin address. Currently it will be a pay-to-address script, but in future (and for other wallets) it could be a script that pays to a segwit address, or a multisig address, or whatever the owner of the wallet wants to be paid to. Have a look at BIP70 for details about payment requests.

Payment requests created via this intent will not contain an amount. The amount is up to the sender.

btw. The second commit in the branch adds an example to the sample app.

**Comment by mahalamobile** [comment]
ok brilliant! So from our side, we call the Intent, if the user doesnt have
your bitcoin wallet what would you suggest?

the example is this right
https://github.com/bitcoin-wallet/bitcoin-wallet/blob/fetch-payment-request/sample-integration-android/src/de/schildbach/wallet/integration/sample/SampleActivity.java

will check BIP70

thanks again

**Comment by schildbach** [comment]
For now, the feature isn't released anyway, so it doesn't matter much what happens if the intent fails because it cannot be resolved. Obviously all testers need know they need to install Bitcoin Wallet alongside your app.

Just before we release the feature, we should try to get other mobile wallets on board. If some of them agree, we could change the intent trigger to something more generic, like an URL. That way, the standard should work with any wallet that has the payment protocol implemented.

**Comment by mahalamobile** [comment]
btw who are the "other" mobile wallets you referring to (see below)? this
is always the problem, too may "vendors" and not enough concencus. Is there
official download and usage stats anywhere so can try and target the top 2
btc android apps? btw a random question, technically if you hold someones
BTc in "cold storage" are you subject to a "banking licence" in effect.

[image: Bitcoin Wallet]Bitcoin
Wallet
[image: Android][image: iOS][image: Windows Phone][image: Linux]
[image: screenshot]
[image: Copay]Copay
[image: Android]
[image: screenshot]
[image: Airbitz Bitcoin Wallet]Airbitz
[image: Android]
[image: screenshot]
[image: Bither]Bither
[image: Android]
[image: screenshot]
[image: GreenBits]GreenBits
[image: Android]
[image: screenshot]
[image: Mycelium]Mycelium
[image: Android][image: iOS]
[image: screenshot]
[image: GreenAddress]Green
Address
[image: Android]
[image: screenshot]
[image: Coinomi Universal Wallet]Coinomi
[image: Android]
[image: screenshot]
[image: Coin.Space]Coin.Space
[image: Android]
[image: screenshot]
[image: Simple Bitcoin Wallet]Simple
Bitcoin

**Comment by schildbach** [comment]
Google Play has some stats, you can go through them yourself. Bitcoin Wallet is the most popular mobile wallet by far, Mycelium comes next and the rest is almost insignificant. I'd try to get a concensus between Mycelium and us, and the others are welcome to join in. Let's keep this discussion to the proposed API.

**Comment by mahalamobile** [comment]
Perfect thanks

Yes agreed.

On Monday, August 15, 2016, Andreas Schildbach <notifications@github.com>
wrote:
> Google Play has some stats, you can go through them yourself. Bitcoin
Wallet is the most popular mobile wallet by far, Mycelium comes next and
the rest is almost insignificant. I'd try to get a concensus between
Mycelium and us, and the others are welcome to join in. Let's keep this
discussion to the proposed API.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci6.googleusercontent.com/proxy/0DqD-RMOHox3dZGSeYuKNSpWUxcJm12x2iLaBPB1THYEQ2InMDjNtWFIClKFynVpkP8qpvB9IT9jfF5Y8mQacBLgNWA2Ikz108cfAouERgHwIH2HINslz98oXbcE738908tG_qoiirXGujLQQf8HaZIZxGHiwg=s0-d-e1-ft#https://github.com/notifications/beacon/AHVQqCYZekROt_v8WQq4huIwk890Y62Oks5qgE9sgaJpZM4DsrZ8.gif>

**Comment by mahalamobile** [comment]
Hi andreas,

Do you know the guys doing the iOS btc wallet?
BIP70 support there, request payment, request address?

Thanks

**Comment by schildbach** [comment]
What do you mean by "the iOS btc wallet"?

**Comment by schildbach** [comment]
For now, this API we're discussing here would be limited to Android. iOS doesn't allow the same degree of inter-app-communication.

**Comment by mahalamobile** [comment]
Hi Andreas,

Are you able to check a payment has arrived via a transactionid from an
external source/app? in otherwords a proof of payment.

Thanks

**Comment by schildbach** [comment]
If you want a safe proof of payment, you need to set up your own full node and query from that. An app that runs on a device owned by an untrusted party cannot be trusted for this purpose.

**Comment by mahalamobile** [comment]
Hi Andreas,

Is there a way in app to change to the testnet blockchain?

thanks

**Comment by mahalamobile** [comment]
Hi Andreas,

So we have managed to get an new receiving address from breadwallet (ios),
however are still struggling to get a new receiving address through android
btc wallet. Seems the fetch payment doesnt expose the new address, we
actually need a "fetchnewaddress" class.

intent.SetComponent(new
Android.Content.ComponentName("de.schildbach.wallet",
"de.schildbach.wallet.ui.fetchnewreceiveaddress");

is this possible?

thanks again
